sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala

```diff
@@ -137,10 +140,14 @@ private[sql] object FileSourceStrategy extends Strategy with Logging {
 
           val splitFiles = selectedPartitions.flatMap { partition =>
             partition.files.flatMap { file =>
+              val blockLocations = getBlockLocations(file)
               (0L to file.getLen by maxSplitBytes).map { offset =>
                 val remaining = file.getLen - offset
                 val size = if (remaining > maxSplitBytes) maxSplitBytes else remaining
-                PartitionedFile(partition.values, file.getPath.toUri.toString, offset, size)
+                // Finds out a list of location hosts where lives the first block that contains the
+                // starting point the file block starts at `offset` with length `size`.
+                val hosts = getBlockHosts(blockLocations, offset)
+                PartitionedFile(partition.values, file.getPath.toUri.toString, offset, size, hosts)
```

Comment:
How is this locality information maintained when we coalesce in the logic below?